### PR TITLE
implement MaskFilter.blur as shadow on Safari

### PR DESCRIPTION
**Pull request by yjbanov**
Fixes https://github.com/flutter/flutter/issues/55930
Requires https://github.com/flutter/goldens/pull/96

### CODE REVIEW

File: lib/web_ui/lib/src/engine/bitmap_canvas.dart [hunk 1]
```diff
@@ -299,50 +300,58 @@ class BitmapCanvas extends EngineCanvas {
 
   @override
   void drawLine(ui.Offset p1, ui.Offset p2, SurfacePaintData paint) {
-    _applyPaint(paint);
-    _canvasPool.strokeLine(p1, p2);
+    _withPaint(paint, () {
```
Comment (ferhatb) [inline lib/web_ui/lib/src/engine/bitmap_canvas.dart:303]: will likely allocate a new class instance for each call due to closure.

File: lib/web_ui/lib/src/engine/vector_math.dart [hunk 2]
```diff
@@ -373,7 +373,9 @@ class Matrix4 {
       return scaled(arg);
     }
     if (arg is Vector3) {
-      return transformed3(arg);
```
Comment (ferhatb) [inline lib/web_ui/lib/src/engine/vector_math.dart:376]: aim was to to reduce allocation/copy, why the change?

File: lib/web_ui/lib/src/engine/canvas_pool.dart [hunk 3]
```diff
@@ -747,10 +748,124 @@ class ContextStateHandle {
...
+  }
+
+  /// Removes paint properties on the current canvas used by the last draw
+  /// command.
+  ///
+  /// Not all properties are cleared. Properties that are set by all paint
+  /// commands prior to painting do not need to be cleared.
+  ///
+  /// Must be called after calling [setUpPaint].
+  void tearDownPaint() {
```
Comment (ferhatb) [inline lib/web_ui/lib/src/engine/canvas_pool.dart:857]: Does this get removed dart2js release? Maybe place all tearDownPaint() calls inside asserts?

File: lib/web_ui/lib/src/engine/bitmap_canvas.dart [hunk 4]
```diff
@@ -478,8 +469,8 @@ class BitmapCanvas extends EngineCanvas {
       if (requiresClipping) {
         restore();
       }
+      _closeCurrentCanvas();
```
Comment (ferhatb) [inline lib/web_ui/lib/src/engine/bitmap_canvas.dart:472]: instead you could remove _closeCurrentCanvas on line 446 in if clause.

File: lib/web_ui/lib/src/engine/bitmap_canvas.dart (hunk 1 above)
Comment (yjbanov) [reply lib/web_ui/lib/src/engine/bitmap_canvas.dart]: Done. Changed it to `_setUpPaint`/`_tearDownPaint` methods.

File: lib/web_ui/lib/src/engine/vector_math.dart (hunk 2 above)
Comment (yjbanov) [reply lib/web_ui/lib/src/engine/vector_math.dart:376]: The copyless version is `transform3`. I removed `transformed3`, which was making copies. I wasn't sure if `operator *` is still used, so I'm not changing it yet.

File: lib/web_ui/lib/src/engine/canvas_pool.dart (hunk 3 above)
Comment (yjbanov) [reply lib/web_ui/lib/src/engine/canvas_pool.dart:857]: It should not be removed. Otherwise the context won't be restored.

File: lib/web_ui/lib/src/engine/bitmap_canvas.dart (hunk 4 above)
Comment (yjbanov) [reply lib/web_ui/lib/src/engine/bitmap_canvas.dart:472]: Hmm, I'm actually not sure what this change is about. Must have been a bad merge 🤔 

Let me check.

File: lib/web_ui/lib/src/engine/canvas_pool.dart (hunk 3 above)
Comment (ferhatb) [reply lib/web_ui/lib/src/engine/canvas_pool.dart:857]: Got it. likely very small will be inlined

File: lib/web_ui/lib/src/engine/bitmap_canvas.dart (hunk 4 above)
Comment (yjbanov) [reply lib/web_ui/lib/src/engine/bitmap_canvas.dart:472]: Ah, right, this was a drive-by change to avoid double-closing. Yes, I will remove the other one. Good catch.

Comment (yjbanov) [reply lib/web_ui/lib/src/engine/bitmap_canvas.dart:472]: Done.

File: lib/web_ui/lib/src/engine/canvas_pool.dart (hunk 3 above)
Comment (yjbanov) [reply lib/web_ui/lib/src/engine/canvas_pool.dart:857]: Would be nice if we had access to dart2js inlining annotations.

**Comment by yjbanov** [comment]
Mac iOS Engine failure is unrelated. Merging.
